services/wfm/src/main/java/org/openkilda/wfm/topology/event/service/PortEventThrottlingService.java

```diff
@@ -0,0 +1,217 @@
+/* Copyright 2018 Telstra Open Source
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.topology.event.service;
+
+import org.openkilda.messaging.info.event.PortChangeType;
+import org.openkilda.messaging.info.event.PortInfoData;
+import org.openkilda.model.SwitchId;
+
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+
+import java.time.Clock;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Slf4j
+public class PortEventThrottlingService {
+    private final Clock clock;
+    private final int minDelay;
+    private final int warmUpDelay;
+    private final int coolDownDelay;
+    private final Map<SwitchPort, PortState> portStats = new HashMap<>();
+
+    /**
+     * The constructor for PortEventThrottlingService.
+     *
+     * @param minDelay the min delay between Port Down / Up change in seconds.
+     * @param warmUpDelay the delay for warm up window in seconds.
+     * @param coolDownDelay the delay for cool down window in seconds.
+     */
+    public PortEventThrottlingService(int minDelay, int warmUpDelay, int coolDownDelay) {
+        this(minDelay, warmUpDelay, coolDownDelay, Clock.systemDefaultZone());
+    }
+
+    /**
+     * This constructor is used only for testing.
+     *
+     * @param minDelay the min delay between Port Down / Up change.
+     * @param warmUpDelay the delay for warm up window.
+     * @param coolDownDelay the delay for cool down window.
+     * @param clock the clock.
+     */
+    PortEventThrottlingService(int minDelay, int warmUpDelay, int coolDownDelay, Clock clock) {
+        this.minDelay = minDelay;
+        this.warmUpDelay = warmUpDelay;
+        this.coolDownDelay = coolDownDelay;
+        this.clock = clock;
+    }
+
+    /**
+     * Processing an event of PortInfoData.
+     * Store it internally for Throttling or return true if Port in UP state and no case is opened for the Port.
+     *
+     * @param data the PortInfoData object.
+     * @return true if the event can be sent further.
+     */
+    public boolean processEvent(PortInfoData data, String correlationId) {
+        SwitchPort switchPort = new SwitchPort(data.getSwitchId(), data.getPortNo());
+        PortState portState = portStats.get(switchPort);
+        if (data.getState() == PortChangeType.UP && portState == null) {
```

Comment:
Should we pass "first" PORT_DOWN without delay too? Event more - we should not start throttling mechanism on PORT_UP event.

It look... natural to pass first PORT_DOWN and it port will continue flapping - drop following UP/DOWN events. When we detect end of flapping - we should emit finite port state.

PS We should not emit 2 PORT_DOWN event in a row.

Reply:
According design - no. We want to hide port flapping. In case when event topology receive `Port Down` and after 300 ms  `Port Up`. The bolt will do nothing, just ignore this.
If it waits `Port Up` for too long, it fire only one `Port Down` and have behavior like you describe above.